internal/argocd/revision.go

```diff
@@ -1,35 +1,82 @@
 // a matching source. If no match is found, an empty string is returned.
 func GetDesiredRevision(app *argocd.Application, freight kargoapi.FreightReference) string {
-	switch {
-	case app == nil || app.Spec.Source == nil:
-		// Without an Application, we can't determine the desired revision.
+	ctx := context.Background()
+	logger := logging.LoggerFromContext(ctx)
+
+	if app == nil {
+		logger.Error(errors.New("Application is nil"), "Cannot determine desired revision for application, bailing.")
 		return ""
-	case app.Spec.Source.Chart != "":
+	}
+
+	appLogger := logger.WithValues("application", app.Name, "namespace", app.Namespace)
+	appLogger.Debug("Getting desired revision for application", "app", app, "spec", app.Spec)
+
+	// An application may have one or more sources.
+	if app.Spec.Source != nil {
+		appLogger.Debug("Application source is not nil, checking.", "source", app.Spec.Source)
+		desiredRevision := GetDesiredRevisionFromSource(ctx, app.Spec.Source, freight)
+		if desiredRevision != "" {
+			return desiredRevision
+		}
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
+	for _, source := range app.Spec.Sources {
+		desiredRevision := GetDesiredRevisionFromSource(ctx, source, freight)
+		if desiredRevision != "" {
+			return desiredRevision
+		}
+	}
+
+	// If we end up here, no desired revision was found from any of the possible sources.
+	appLogger.Debug("Could not determine desired revision for application from any sources.")
+	return ""
+}
+
+func GetDesiredRevisionFromSource(
+	ctx context.Context,
+	s argocd.ApplicationSource,
+	freight kargoapi.FreightReference) string {
+
+	logger := logging.LoggerFromContext(ctx)
+	sourceLogger := logger.WithValues("source", s)
+	sourceLogger.Debug("Getting desired revision for application source", "source", s)
+
+	switch {
+	case s.Chart != "":
 		// This source points to a Helm chart.
 		// NB: This has to go first, as the repository URL can also point to
 		//     a Helm repository.
-		sourceChart := path.Join(app.Spec.Source.RepoURL, app.Spec.Source.Chart)
+		sourceChart := path.Join(s.RepoURL, s.Chart)
+		sourceLogger.Debug("Application source is a Helm chart", "repoURL", s.RepoURL, "chart", s.Chart)
 		for _, chart := range freight.Charts {
 			// Join accounts for the possibility that chart.Name is empty.
 			if path.Join(chart.RepoURL, chart.Name) == sourceChart {
 				return chart.Version
 			}
 		}
-	case app.Spec.Source.RepoURL != "":
+	case s.RepoURL != "":
 		// This source points to a Git repository.
-		sourceGitRepoURL := git.NormalizeURL(app.Spec.Source.RepoURL)
+		sourceLogger.Debug("Application source is a Helm chart, without a chart name", "repoURL", s.RepoURL, "path", s.Path)
```

Comment:
Is this true? I was under the impression that at this point it's either a chart from an HTTP repository OR a Git repository revision.

Reply:
Yes, which means the current log message does not take it potentially being a Git repository into account?

Reply:
My bad, I realize I misunderstood your question, but yes, you can have a Helm chart source with a Path instead of a chart name.

Reply:
We still are not aligned.

My point is that the log message is incorrect... Any Helm chart scenario is handled by the `case s.Chart != ""`. While at the point where you added the log message, it will start looking for commit references.

It should thus instead be something like:

```suggestion
sourceLogger.Debug("Application source is a Git repository", "repoURL", s.RepoURL, "path", s.Path)
```

Reply:
That makes sense, updated.